tests/unittest.cpp

```diff
@@ -411,6 +420,130 @@ TEST_CASE("const correct iterators on collections", "[const-correctness]") {
 }
 
+TEST_CASE("Iterator concepts", "[iterator-concepts]") {
+  // a collection to play with
+  auto collection = ExampleHitCollection();
+
+  // bidirectional iterator traits
+  STATIC_REQUIRE(std::is_same_v<std::iterator_traits<decltype(collection)::iterator>::iterator_category, std::bidirectional_iterator_tag>);
+  STATIC_REQUIRE(std::is_same_v<std::iterator_traits<decltype(collection)::const_iterator>::iterator_category, std::bidirectional_iterator_tag>);
+  STATIC_REQUIRE(std::is_same_v<std::iterator_traits<decltype(collection)::reverse_iterator>::iterator_category, std::bidirectional_iterator_tag>);
+  STATIC_REQUIRE(std::is_same_v<std::iterator_traits<decltype(collection)::const_reverse_iterator>::iterator_category, std::bidirectional_iterator_tag>);
+
+  #ifdef __cpp_lib_ranges
+  // bidirectional iterator concepts (C++20)
+  STATIC_REQUIRE(std::bidirectional_iterator<decltype(collection)::iterator>);
+  STATIC_REQUIRE(std::bidirectional_iterator<decltype(collection)::const_iterator>);
+  STATIC_REQUIRE(std::bidirectional_iterator<decltype(collection)::reverse_iterator>);
+  STATIC_REQUIRE(std::bidirectional_iterator<decltype(collection)::const_reverse_iterator>);
+  STATIC_REQUIRE(std::ranges::bidirectional_range<decltype(collection)>);
+  #endif
+
+  // iterator loops
+  auto cellID = 0;
+  collection.clear();
+  collection.create();
+  collection.create(0x42ULL, 0., 0., 0., 0.);
+  // forward loop with post-increment
+  for (auto c = collection.begin(); c != collection.end(); c++) {
+    STATIC_REQUIRE(std::is_same_v<decltype(c), ExampleHitMutableCollectionIterator>);
+    cellID = c->cellID();
+  }
+  REQUIRE(cellID == 0x42ULL);
+  // reverse const loop with pre-increment
+  for (auto c = collection.crbegin(); c != collection.crend(); ++c) {
+    STATIC_REQUIRE(std::is_same_v<decltype(c), ExampleHitCollectionReverseIterator>);
+    cellID = c->cellID();
+  }
+  REQUIRE(cellID == 0);
+
+  // fill from const
+  collection.clear();
+  collection.create();
+  auto const_hit = ExampleHit(0x42ULL, 0., 0., 0., 0.);
+  //FIXME this should work but we may need a conversion constructor!
+  //std::fill(collection.begin(), collection.end(), const_hit);
+  //REQUIRE(collection.begin()->cellID() == 0x42ULL);
+  // fill from mutable
+  collection.clear();
+  collection.create();
+  auto mutable_hit = MutableExampleHit(0x42ULL, 0., 0., 0., 0.);
+  std::fill(collection.begin(), collection.end(), mutable_hit);
+  //FIXME
+  //REQUIRE(collection.begin()->cellID() == 0x42ULL);
+
+  // find in collection
+  collection.clear();
+  collection.create();
+  collection.create(0x42ULL, 0., 0., 0., 0.);
+  auto found = std::find_if(collection.begin(), collection.begin(),
+    [](const auto& h){ return h.cellID() == 0x42ULL; });
+  STATIC_REQUIRE(std::is_same_v<decltype(found), ExampleHitMutableCollectionIterator>);
+  REQUIRE(found != collection.end());
+  //FIXME
+  //REQUIRE(found->cellID() == 0x42ULL);
```

Comment:
Is this still commented because it is not working yet, or is this leftover from previous iterations / developing?

Reply:
This one does still need fixing indeed. Not sure yet what's going on with this.

Reply:
`std::find_if` was just a stupid mistake (which took me way too long to catch) where the search boundaries were both `begin()`...